### PR TITLE
Added Config for custom driver (specifically libSQL)

### PR DESCRIPTION
**Pull request by ytsruh**
- [x] Do only one thing
- [x] Non breaking API changes
- [x] Tested

### What did this pull request do?

It adds a Config struct that can be used the same way the Postgres version can be. This allows a user to add their own driver instead of using the pre-defined one.

### User Case Description
To use this with libSQL or Turso compatible SQLite

### CODE REVIEW

File: sqlite.go [hunk 1]
```diff
@@ -24,21 +24,30 @@ type Dialector struct {
...
 
 func (dialector Dialector) Initialize(db *gorm.DB) (err error) {
-	if dialector.DriverName == "" {
-		dialector.DriverName = DriverName
-	}
 
 	if dialector.Conn != nil {
 		db.ConnPool = dialector.Conn
+	} else if dialector.DriverName != "" {
+		db.ConnPool, err = sql.Open(dialector.DriverName, dialector.DSN)
```
[the remark this answers is not in the document]
Comment (ytsruh) [reply sqlite.go:50]: Erm, yes! In my (very poor) defence I adapted the approach from the Postgres version and its actually missing there too. Will resolve tonight.

I'll have to look at these tests failing as well as I cannot figure out why

File: sqlite.go [hunk 2]
```diff
@@ -24,21 +24,30 @@ type Dialector struct {
...
 }
 
 func (dialector Dialector) Initialize(db *gorm.DB) (err error) {
-	if dialector.DriverName == "" {
-		dialector.DriverName = DriverName
-	}
 
 	if dialector.Conn != nil {
 		db.ConnPool = dialector.Conn
+	} else if dialector.DriverName != "" {
```
Comment (haaawk) [inline sqlite.go:49]: Is this if branch doing anything different than the else brach below?

Comment (ytsruh) [reply sqlite.go]: No. I've removed it. I was hugely overengineering the solution to this problem!

File: sqlite.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ytsruh) [reply sqlite.go]: It's been removed. It wasn't required as mentioned in the other comment and the error handling is now not required.

**Review by haaawk — APPROVED** [review]
LGTM

**Review by dangtran47 — APPROVED** [review]

**Comment by asilvadesigns** [comment]
👀

**Comment by xEricL** [comment]
Looking forward to this! 😃

**Comment by xEricL** [comment]
Just to confirm, I tested the changes in this PR and everything works flawlessly with Turso. Nice work @ytsruh!

**Comment by haaawk** [comment]
What is needed to merge this?
@jinzhu Could you have a look please?

**Comment by ncruces** [comment]
I'll be blunt: this is a _bad_ idea.

The API is fine, although I'm partial to just adding a `func OpenConn(conn gorm.ConnPool)` and be done with it.

The reason _why_ this is a bad idea is the intended purpose of using alternative drivers. I cannot stress this enough: using multiple implementations of SQLite in the same process is a really, _really_ bad idea. If you, by accident, open the same database with different drivers, the most likely result is **database corruption** [^1].

As long as this package imports `github.com/mattn/go-sqlite3` we shouldn't encourage using it with any other drivers.

You can easily make a copy of this, and publish your own Turso/libSQL Gorm driver. Same as was done by @glebarez for `modernc.org/sqlite`.

If we _want_ to encourage alternative drivers, this package should stop importing `github.com/mattn/go-sqlite3` and simply assume the driver is registered elsewhere.

[^1]: the [POSIX locking mechanisms](https://sqlite.org/src/artifact/2e8b12?ln=1073-1161) SQLite depends on _require_ global variables and if you bundle multiple versions of the library, you'll have multiple copies of those global variables.

**Comment by haaawk** [comment]
> I'll be blunt: this is a _bad_ idea.
> 
> The API is fine, although I'm partial to just adding a `func OpenConn(conn gorm.ConnPool)` and be done with it.
> 
> The reason _why_ this is a bad idea is the intended purpose of using alternative drivers. I cannot stress this enough: using multiple implementations of SQLite in the same process is a really, _really_ bad idea. If you, by accident, open the same database with different drivers, the most likely result is **database corruption** [1](#user-content-fn-1-75ab97e6766ecc4c22b3324a830e89e9).
> 
> As long as this package imports `github.com/mattn/go-sqlite3` we shouldn't encourage using it with any other drivers.
> 
> You can easily make a copy of this, and publish your own Turso/libSQL Gorm driver. Same as was done by @glebarez for `modernc.org/sqlite`.
> 
> If we _want_ to encourage alternative drivers, this package should stop importing `github.com/mattn/go-sqlite3` and simply assume the driver is registered elsewhere.
> 
> ## Footnotes
> 1. the [POSIX locking mechanisms](https://sqlite.org/src/artifact/2e8b12?ln=1073-1161) SQLite depends on _require_ global variables and if you bundle multiple versions of the library, you'll have multiple copies of those global variables. [↩](#user-content-fnref-1-75ab97e6766ecc4c22b3324a830e89e9)

I'm not sure this is a problem at all. Neither `https://github.com/tursodatabase/libsql-client-go` nor `https://github.com/tursodatabase/go-libsql` import
`github.com/mattn/go-sqlite3` or `modernc.org/sqlite`. Turso uses http to access remote instance of sqlite.

**Comment by haaawk** [comment]
If you decide to do your own "clone" @ytsruh you can take inspiration in https://github.com/ncruces/go-sqlite3/blob/main/gormlite/sqlite.go

**Comment by ncruces** [comment]
> I'm not sure this is a problem at all. Neither `github.com/tursodatabase/libsql-client-go` nor `github.com/tursodatabase/go-libsql` imports `github.com/mattn/go-sqlite3` or `modernc.org/sqlite`. Turso uses http to access remote instance of sqlite.

Then maybe it isn't a problem for Turso. But it is a problem to encourage this kind of flexibility without a big disclaimer in the documentation.

Just mixing `mattn`/`modernc` is enough to corrupt a database. And it's particularly bad that they chose non-conflicting driver names for this exact reason, but that was `modernc`'s decision.

**Comment by ncruces** [comment]
> If you decide to do your own "clone" @ytsruh you can take inspiration in https://github.com/ncruces/go-sqlite3/blob/main/gormlite/sqlite.go

It's very similar to this one, as is @glebarez. The only real difference is the error handling.

The situation with locking is very unfortunate, but people working on SQLite bindings and drivers need to put making it hard for users to corrupt data by accident at the very top.

**Comment by ytsruh** [comment]
Hmm I would rather get some feedback from the maintainer. The idea to use the config struct originally came from Jinzhu (in another thread). I would rather it be part of the original package but I dont want any breaking changes. 

I'll give it another week and see if any feedback comes. If not, I'll clone my own and create some docs.

**Comment by ytsruh** [comment]
Hi all, I've decided to go solo/indie on this one. Package should now be available at https://github.com/ytsruh/gorm-libsql

```go get -u github.com/ytsruh/gorm-libsql```

I'd still prefer this to be part of the gorm library so if it is merged in then I willa archive my package and point any users back to Gorm. Until that happens, its useable (with docs in the readme).

Thanks everybody for your support/feedback on this

**Comment by haaawk** [comment]
Thanks for doing this @ytsruh !!!

**Comment by xEricL** [comment]
> Hi all, I've decided to go solo/indie on this one. Package should now be available at https://github.com/ytsruh/gorm-libsql
> 
> `go get -u github.com/ytsruh/gorm-libsql`
> 
> I'd still prefer this to be part of the gorm library so if it is merged in then I willa archive my package and point any users back to Gorm. Until that happens, its useable (with docs in the readme).
> 
> Thanks everybody for your support/feedback on this

@ytsruh When I run `go get -u github.com/ytsruh/gorm-libsql` I get the error
```
go: downloading github.com/ytsruh/gorm-libsql v0.1.1
go: github.com/ytsruh/gorm-libsql@upgrade (v0.1.1) requires github.com/ytsruh/gorm-libsql@v0.1.1: parsing go.mod:
        module declares its path as: gorm.io/driver/sqlite
                but was required as: github.com/ytsruh/gorm-libsql
```

Also, in your [README.md](https://github.com/ytsruh/gorm-libsql/blob/c32a62ea7485bcb5672da48172ba41dc35005412/README.md?plain=1#L23), your import uses "https://":
```
sqlite "https://github.com/ytsruh/gorm-libsql"
```

**Comment by ytsruh** [comment]
Sorry about that! (Both of them) Both fixed now...

... https://pkg.go.dev/github.com/ytsruh/gorm-libsql

**Comment by jinzhu** [comment]
Sorry for the delay. This PR looks good, so I'll merge it first. If you want your new driver to be added to the GORM documentation, you can submit a PR to gorm.io.

**Comment by ncruces** [comment]
This PR is just depressing.

Where does [libsql-client-go](https://github.com/tursodatabase/libsql-client-go) even return an error that's compatible with [error_translator.go](https://github.com/go-gorm/sqlite/blob/master/error_translator.go)?

@glebarez has its own [error translation](https://github.com/glebarez/sqlite/blob/2051f80732bf378c7cd16d7325819d73823dd481/sqlite.go#L238-L251), gormlite does the [same](https://github.com/ncruces/go-sqlite3/blob/main/gormlite/error_translator.go).

This just does... nothing?

It's like #161: it just ignores the issues, and acts like they don't exist.

Do the [extensive Gorm unit tests](https://github.com/glebarez/sqlite?tab=readme-ov-file#is-this-tested-good-) even pass? No one knows, no one cares, apparently.

**Comment by avinassh** [comment]
TIL about GORM tests. May be they should be part of this repo's CI? In any case, those can be worked and fixed.

**Comment by ncruces** [comment]
They are part of GORM's CI.

This GORM driver is tested **with the mattn SQLite driver** on GORM's CI.

If you want to use another SQLite driver with this GORM driver, you should run tests, like:
- @glebarez: https://github.com/glebarez/sqlite/blob/master/.github/workflows/gorm-tests.yml
- gormlite: https://github.com/ncruces/go-sqlite3/blob/main/gormlite/test.sh

**Comment by xEricL** [comment]
@jinzhu what are your thoughts on @ncruces's concerns?

**Pull request merged**